### ISSUE TITLE
Add and remove card types triggers a full sync even if user discards changes

### ISSUE THREAD

**Issue by sudomain** [ankidroid/Anki-Android#6475]
Something I noticed while poking around CardTemplateEditor.java and TemporaryModel.java:
###### Reproduction Steps

1. Add or remove a card type
2. Agree to the full sync prompt
3. Back out and discard changes
4. Sync

###### Expected Result
Since the changes (either add or remove) were discarded, a full sync shouldn't be needed

###### Actual Result
Full sync is required

###### Debug info
Refer to the [support page](https://ankidroid.org/docs/help.html) if you are unsure where to get the "debug info".

AnkiDroid Version = 2.12alpha3

Android Version = 7.0

ACRA UUID = 0f431910-fac3-4b52-ba37-64f328e5d239

###### Research
*Enter an [x] character to confirm the points below:*

- [x ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [x ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x ] I have searched for similar existing issues here and on the user forum
- [x ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

**Comment by mikehardy** [comment]
Unfortunately, once you trigger that I don't think it's possible to un-trigger it, but it does bear looking at, maybe it is possible to un-trigger that if the user doesn't save changes and I just missed it

**Comment by mikehardy** [comment]
Incidentally, all the usage you are doing on CardTemplateEditor is much appreciated, it's not the most heavily used area of the code but it just saw so much change that more testing == more better :-)

**Comment by sudomain** [comment]
> Unfortunately, once you trigger that I don't think it's possible to un-trigger it, but it does bear looking at, maybe it is possible to un-trigger that if the user doesn't save changes and I just missed it

I haven't looked into too much, but perhaps we can move the `try` / `catch` block and `mTemplateEditor.getCol().modSchema();` line into the function that saves the model. Perhaps a separate prompt when adding/removing a card type saying "If you save this note type after adding or removing a card type, a full collection sync will be required" with an `Ok` button. I probably didn't explain this well, but I'll try it and get back to you.

**Comment by mikehardy** [comment]
No - that was discussed quite a bit during development of 5151. The prompt for full sync needs to be associated-in-time with the action the user does that requires it, even though we don't actually do the save right then. That's inline with desktop actually.

**Comment by timrae** [comment]
I don't think full sync is undoable on Anki Desktop either is it...?

**Comment by sudomain** [comment]
I don't think desktop collects changes made on the template screen, thus once an add/delete template change is made the only way of undoing it is to download from ankiweb. It looks like this behavior is planned for desktop though (quoting from the [2.1.28alpha1 release notes)](https://betas.ankiweb.net/#/?id=changes-in-2128alpha1):

`Card template screen:

    Changes are now accumulated, and can be saved or discarded when you close the screen.`

I don't use the desktop alphas, so I can't comment.

**Comment by timrae** [comment]
True, could be something we get 'for free' by switching to rust then

**Comment by david-allison** [comment]
@sudomain this should be fixed in 2.17, could you confirm when you have a second

**Comment by sudomain** [comment]
I reproduced it just now on 2.17beta3

**Comment by david-allison** [comment]
* Superseded by https://github.com/ankidroid/Anki-Android/issues/19956
